File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.py
class Solution:
    def maxLength(self, arr: List[str]) -> int:
        n = len(arr)
        self.ans = 0
        
        def checkUnique(path):
            ms = set()
            for char in path:
                if char in ms:
                    return False
                ms.add(char)
            return True
        
        def backTracking(i, path):
            if not checkUnique(path):
                return
            
            if len(path) > self.ans:
                self.ans = len(path)
            
            for j in range(i, n):
                backTracking(j + 1, path + arr[j])
                
        backTracking(0, '')
        return self.ans
                
            

File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    int maxLength(vector<string>& arr) {
        int n = arr.size();
        int ans = 0;
        
        for (int mask = 0; mask < (1 << n); ++mask) {
            int seen = 0, len = 0;
            bool flag = false;
            for (int i = 0; i < n; ++i) {
                if (((mask >> i) & 1) == 0) {
                    continue;
                }
                for (char c : arr[i]) {
                    int j = c - 'a';
                    if ((seen >> j) & 1) {
                        flag = true;
                        break;
                    }
                    seen |= (1 << j);
                    len += 1;
                }
                if (flag) {
                    break;
                }
            }
            if (!flag and len > ans) {
                ans = len;
            }
        }
        
        return ans;
    }
};